[zstd_stream] Don't block in reader.Read if a zstd block is available

delthas opened the pull request:
reader.Read used to try to fully read an internal buffer until EOF or
the buffer was filled. That was buffer was set to ZSTD_DStreamInSize,
which is larger than any zstd block.

This means that reader.Read could try to buffer much more data than
what was needed to process and return a single block from the Read
method.

This was an issue because we could miss an urgent Flush from a
corresponding Writer by blocking. (A typical use case is instant
messaging.) It was also against the general convention of io.Read that a
single call should return any immediately available data without
blocking, if any.

Interestingly enough, the test case should have caught this up, but
because we used a bytes.Buffer, the Read returned io.EOF after reading
the entirety of the buffer, even if we appended to the buffer later on.
The test case is also fixed by this commit.

Fixes: #95

delthas commented:
It could be interesting to run some benchmarks to see if that impacts performance, and if yes, to take a decision on whether that should be the default or flag-dependent.

delthas commented:
Benchmarks:
Before *(modified the Decompression test to use io.ReadAll so that we test the same stuff as After)*:
```
goos: windows
goarch: amd64
pkg: github.com/DataDog/zstd
cpu: Intel(R) Core(TM) i7-3820 CPU @ 3.60GHz
BenchmarkStreamCompression
BenchmarkStreamCompression-8     	     248	 143712750 ns/op	  69.38 MB/s
BenchmarkStreamDecompression
BenchmarkStreamDecompression-8   	    2014	  17949798 ns/op	 555.47 MB/s
PASS
```

After:
```
goos: windows
goarch: amd64
pkg: github.com/DataDog/zstd
cpu: Intel(R) Core(TM) i7-3820 CPU @ 3.60GHz
BenchmarkStreamCompression
BenchmarkStreamCompression-8     	     244	 144691207 ns/op	  68.91 MB/s
BenchmarkStreamDecompression
BenchmarkStreamDecompression-8   	    1989	  18085000 ns/op	 551.32 MB/s
PASS
```

No significant change for compression (as expected), ~0.7% slowdown for decompression. IMO no need to add a flag for that, the performance difference is OK.

delthas commented:
Ready for review @Viq111

delthas commented:
Gentle ping 😃

Viq111 commented:
Hi! Sorry I dropped the ball on this one so that's totally on me.
I was out last week but I'm back this week so I'll be able to give it a look.

Viq111 commented on zstd_stream.go:
```diff
@@ -422,29 +422,53 @@ func (r *reader) Read(p []byte) (int, error) {
+				return 0, fmt.Errorf("failed to read from underlying reader: %s", err)
+			}
+			if n == 0 {
+				return 0, io.EOF
```
I'm initially thinking of changing the condition to:
```go
remainingOut := r.decompSize - r.decompOff
if err == io.EOF && r.compressionLeft == 0 && remainingOut == 0 {
```

delthas replied:
My understanding is that if `r.decompSize < len(r.decompressionBuffer)`, ie if `output.pos < output.size`, the decoder flushed all it could (see https://github.com/DataDog/zstd/blob/1.x/zstd.h#L774); in other words the decoder needs more data in order to output anything else.

The line you linked seems to refer to a special condition if `len(r.decompressionBuffer) > ZSTD_BLOCKSIZE_MAX`. However this will never be the case, because the decompression buffer is allocated by a pool to a size of `ZSTD_DStreamOutSize`, which is == `ZSTD_BLOCKSIZE_MAX`. (Which is the purpose of using `ZSTD_DStreamOutSize()`.)

So the condition looks correct to me as is. What do you think?

Viq111 replied:
Indeed! I didn't read on `ZSTD_DStreamOutSize == ZSTD_BLOCKSIZE_MAX` but yes you are correct

Viq111 commented on zstd_stream.go:
```diff
@@ -489,25 +512,9 @@ func (r *reader) Read(p []byte) (int, error) {
-		err = errShortRead
+		if r.decompOff > 0 {
+			return r.decompOff, nil
+		}
```
Thinking also about a input stream that got accidentaly cut off (so it starts returning `EOF` but we still have some zstd partial data), we could return a `io.UnexpectedEOF`
https://golang.org/pkg/io/#pkg-variables
> Functions should return EOF only to signal a graceful end of input. If the EOF occurs unexpectedly in a structured data stream, the appropriate error is either ErrUnexpectedEOF or some other error giving more detail.

delthas replied:
Added some code & comment about that. https://github.com/DataDog/zstd/pull/96/files#diff-ff3e40bda515a4af0936733c5415920eb09cc903913a7243bdec1074266784b0R470  Thoughts?

Viq111 reviewed (commented):
Thanks for the PR! It definitely makes a lot of sense
Rereading the stream decompression path, I think we can handle / document a bit better all the `EOF` cases

Viq111 commented:
For benchmarking, I think those changes are immaterial so I would not worry about them:
```
ᐅ benchstat
goos: darwin
goarch: amd64
pkg: github.com/DataDog/zstd
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
name                    old time/op    new time/op    delta
CtxCompression-16         46.8µs ± 1%    50.5µs ±13%   ~     (p=0.065 n=9+10)
CtxDecompression-16       2.70µs ± 5%    2.95µs ±15%   ~     (p=0.247 n=10+10)
StreamCompression-16       869ns ±13%     898ns ±12%   ~     (p=0.190 n=10+10)
StreamDecompression-16    2.50µs ± 7%    2.57µs ±14%   ~     (p=0.952 n=10+10)
Compression-16            51.0µs ± 8%    52.1µs ± 8%   ~     (p=0.684 n=10+10)
Decompression-16          3.07µs ±16%    3.11µs ±15%   ~     (p=0.542 n=10+10)

name                    old speed      new speed      delta
CtxCompression-16        151MB/s ± 1%   140MB/s ±12%   ~     (p=0.065 n=9+10)
CtxDecompression-16     2.61GB/s ± 5%  2.42GB/s ±14%   ~     (p=0.247 n=10+10)
StreamCompression-16    8.15GB/s ±12%  7.90GB/s ±11%   ~     (p=0.190 n=10+10)
StreamDecompression-16  2.83GB/s ± 7%  2.76GB/s ±13%   ~     (p=0.971 n=10+10)
Compression-16           139MB/s ± 7%   136MB/s ± 8%   ~     (p=0.684 n=10+10)
Decompression-16        2.31GB/s ±14%  2.28GB/s ±13%   ~     (p=0.529 n=10+10)
```

delthas committed:
[zstd_stream] Don't block in reader.Read if a zstd block is available

reader.Read used to try to fully read an internal buffer until EOF or
the buffer was filled. That was buffer was set to ZSTD_DStreamInSize,
which is larger than any zstd block.

This means that reader.Read could try to buffer much more data than
what was needed to process and return a single block from the Read
method.

This was an issue because we could miss an urgent Flush from a
corresponding Writer by blocking. (A typical use case is instant
messaging.) It was also against the general convention of io.Read that a
single call should return any immediately available data without
blocking, if any.

Interestingly enough, the test case should have caught this up, but
because we used a bytes.Buffer, the Read returned io.EOF after reading
the entirety of the buffer, even if we appended to the buffer later on.
The test case is also fixed by this commit.

Fixes: #95

delthas commented:
Gentle ping :grinning:

Viq111 committed:
Merge branch '1.x' into fix-stream-read
